src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -587,6 +612,93 @@ StoreQueryResult(PGconn *connection, TupleDesc tupleDescriptor,
 
 
 /*
+ * ConsumeQueryResult gets a query result from a connection, counting the rows
+ * and checking for errors, but otherwise discarding potentially returned
+ * rows.  Returns true if a non-error result has been returned, false if there
+ * has been an error.
+ */
+static bool
+ConsumeQueryResult(PGconn *connection, int64 *rows)
+{
+	bool gotResponse = false;
+
+	*rows = 0;
+
+	/*
+	 * Due to single row mode we have to do multiple PQgetResult() to finish
+	 * processing of this query, even without RETURNING. For single-row mode
+	 * we have to loop until all rows are consumed.
+	 */
+	while (true)
+	{
+		PGresult *result = PQgetResult(connection);
+		ExecStatusType status = PGRES_COMMAND_OK;
+
+		if (result == NULL)
+		{
+			break;
+		}
+
+		status = PQresultStatus(result);
+
+		if (status != PGRES_COMMAND_OK &&
+			status != PGRES_SINGLE_TUPLE &&
+			status != PGRES_TUPLES_OK)
+		{
+			char *sqlStateString = PQresultErrorField(result, PG_DIAG_SQLSTATE);
+			int category = 0;
+			bool raiseError = false;
+
+			/*
+			 * If the error code is in constraint violation class, we want to
+			 * fail fast because we must get the same error from all shard
+			 * placements.
+			 */
+			category = ERRCODE_TO_CATEGORY(ERRCODE_INTEGRITY_CONSTRAINT_VIOLATION);
+			raiseError = SqlStateMatchesCategory(sqlStateString, category);
```

Comment:
While I was doing some failure testing, I realized that `sqlStateString` becomes empty when two nodes are down but the connections have already been cached (i.e., status == `PGRES_FATAL_ERROR`). This makes `SqlStateMatchesCategory()` to crash.

Steps to reproduce:
```SQL

CREATE TABLE multiple_hash 
( 
category text NOT NULL PRIMARY KEY, 
data text NOT NULL
);

SELECT master_create_distributed_table('multiple_hash', 'category', 'hash');
SELECT master_create_worker_shards('multiple_hash', 4, 2);

INSERT INTO multiple_hash VALUES ('1', '1');
```

```
-- now terminate both worker nodes nodes
/usr/local/pgsql/bin/pg_ctl -D /Users/onderkalaci/Documents/data_dir/worker_9700 -o "-p 9700" -l   /tmp/logfile_9700 stop -m f

/usr/local/pgsql/bin/pg_ctl -D /Users/onderkalaci/Documents/data_dir/worker_9701 -o "-p 9701" -l   /tmp/logfile_9701 stop -m f
```

```SQL
INSERT INTO multiple_hash VALUES ('1', '1');
WARNING:  terminating connection due to administrator command
CONTEXT:  while executing command on localhost:9701
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
```